doc/contributing.rst

```diff
@@ -671,3 +671,118 @@ Conventions
    Furthermore, module names should use `snake_case` and not use capital
    letters, which cause issues when going from an OS without case
    sensitivity to an OS with it.
+
+
+Breaking Changes
+================
+
+Introducing breaking changes, no matter how well intentioned,
+creates long-term problems for the community, in particular those looking to protomote
+reusable Nim code in libraries: In the Nim distribution, critical security and bugfixes,
+language changes and community improvements are bundled in a single distribution - it is
+difficult to make partial upgrades with only benign changes. When one library depends on
+a legacy behavior, it can no longer be used together with another library that does not,
+breaking all downstream applications - the standard library is unique in that it sits at
+the root of **all** dependency trees.
+
+There is a big difference between compile-time breaking changes and run-time breaking
+changes.
+
+
+Run-time breaking changes
+-------------------------
+
+Run-time breaking changes are to be avoided at almost all costs: Nim is used for
+mission critical applications which depend on behaviours that
+are not covered by the test suite. As such, it's important that changes to the
+*stable* parts of the standard library are made avoiding changing the existing
+behaviors, even when the test suite continues to pass.
+
+Examples of run-time breaking changes:
+
+- Raising exceptions of a new type, compared to what's currently being raised.
```

Comment:
Raising a Defect should be considered an acceptable breaking change when the defect is called for

Reply:
Defect is an annoying things about Nim.

The last thing I need is some library author thinking they know better than I do in regards to what is recoverable. Which is rarely if ever possible for the author to know as per the end to end argument in system design, the consumer of the library almost always has more options at their disposal and an uncatchable exception really stymes that.

Reply:
I don't understand. So what do you propose?

Reply:
> Raising a Defect should be considered an acceptable breaking change when the defect is called for

Hardly, as it's a subtle runtime change.

Reply:
Raising an exception of a new type is a lesser breaking change than raising a defect, the latter should be scrutinized further.

Anyhow, as it's written is better than what @timotheecour is proposing.

Reply:
just an example, but see what dotnet allows/disallows: https://docs.microsoft.com/en-us/dotnet/core/compatibility/ agrees
> ✔️ ALLOWED: Throwing an exception that is considered unrecoverable

> ✔️ ALLOWED: Throwing a new exception in a new code path

> The exception must apply only to a new code-path that's executed with new parameter values or state and that can't be executed by existing code that targets the previous version.

Reply:
"in a new code path" is not clear enough and is better left out.

Reply:
the meaning is: if the new exception can only be raised when new optional params are explicitly provided, it's an acceptable breaking change:
```nim
# before:
proc fn*(a: int) = discard
# after v1:
proc fn*(a: int, dir = "") =
  assert dir.len == 0 or dir.isAbsolute # with a Defect
```
existing code that uses fn will not break

# after v2:
proc fn*(a: int, dir = "") =
  if dir != "" and not dir.isAbsolute: raise newException(ValueError, dir) # with a ValueError
```
existing code that uses fn will not raise, at worse you get a CT error if user has `raises: []` on a caller of `fn` (falls under category of CT breaking change, not RT breaking change)

Reply:
Sounds reasonable, but this rule can added later.